cirq/devices/hypergraph.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import (
+        Any, Dict, FrozenSet, Hashable, Iterable, Optional, Set, Tuple)
+
+
+AdjacencyList = Set[FrozenSet[Hashable]]
+
+class UndirectedHypergraph:
```

Comment:
Does `networkx` include a class for this already? Ideally we wouldn't be defining our own graph data structures.

Reply:
Unfortunately not. I was tempted to use networks and restricted to (non-hyper-)graphs, but I think that could be overly limiting down the road. Two expected use cases for hyperedges are ion traps (which may have native more-than-two-qubit gates) and abstractions of superconducting devices that assume a "native" four-qubit gate (e.g. when compiling quantum chemistry circuits).